### ISSUE TITLE
Shard engine build

### ISSUE THREAD

**Issue by cbracken** [flutter/flutter#26730]
Once we've completed migration of our ChromeBot build to LUCI, we should investigate sharding the build on (host_os, target_embedder, runtime_mode) where:

  * host_os = { Linux, Mac, Windows }
  * target_embedder = { android-armv7, android-arm64, iOS } (the embedder API .so is currently built as part of the host build)
  * runtime_mode = { debug_unopt, profile, dynamic_profile, release, dynamic_release }

Note that while iOS has armv7 + arm64 support, we always build both and merge into a universal binary. On Android, we can build these entirely separately.

If we had to sacrifice something here, sharding 3 ways and doing profile+dynamic_profile as part of the same build, and release+dynamic_release as part of the same build seems the most obvious sacrifice to make.

We'll need to refactor how we do our upload to cloud storage and make corresponding changes in flutter_tools. As one example, today we bundle all the android-arm64 build products together in a single zip (across runtime modes).

**Comment by liyuqian** [comment]
@cbracken : for the auto-roller triggering question, we actually don't (know a way to programmatically) check whether all builds are green or not. Instead, we check if the Google cloud bucket has the artifacts compiled for a specific engine version. It's possible that the build is red but the roller will trigger since it finds all the artifacts in the check list. In that case, we usually then modify our check list to include more artifacts. So far, there's no guarantee that the auto-roller won't be triggered if the build is red.

**Comment by dnfield** [comment]
I believe LUCI can send notifications when the build is completed

**Comment by liyuqian** [comment]
CC @rmistry to see if auto-roller can utilize the LUCI notifications.

**Comment by rmistry** [comment]
The roller does not have support for LUCI notifications (I did not know they existed). A new repo manager (in the roller code) was written to wait for when the bucket artifacts were available. Something new would have to be written for doing something similar for LUCI builds. It would be a fair bit of work and would not happen anytime soon.
The rollers do have support for checking on and waiting for "flutter-build". Can that check be modified to include the status of these builds as well? or maybe another similar check can be created.

**Comment by cbracken** [comment]
@chinmaygarde @zanderso thoughts on closing this as obsolete? We've done a few years of sharding since this was filed, and I don't know that there's anything in this bug that we either haven't done or aren't planning to do :)

**Comment by godofredoc** [comment]
The plan is to close this bug once we complete the migration to engine v2 recipes.

**Comment by zanderso** [comment]
@godofredoc is there another issue for the engine v2 recipes that we can x-ref here?

**Comment by godofredoc** [comment]
Yes, this is the one where we have been tracking the progress with engine v2 recipes: https://github.com/flutter/flutter/issues/81855

**Comment by godofredoc** [comment]
This is now complete with all the engine_v2 builds running on prod.
